REGR: fix return class in _constructor_from_mgr for simple subclasses

ivirshup opened the pull request:
- [x] closes #55763 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This is a draft for fixing #55763. I'm not familiar with this part of the code, but this seems to fix the issue locally.

This is addressing changes brought in by https://github.com/pandas-dev/pandas/pull/54922. Instead of checking the class of `._constructor` in `_constructor_from_mgr`, this checks the class being called.

We were relying on most operations on our subclass returning a pandas dataframe. AFAICT, the previous PR changed the behavior so that most operations would return the class of the instance even if `_constructor` was not set.


Once I see that this doesn't break tests on CI, added tests will be like:

```python
class SimpleSubClass(DataFrame):
    """A subclass of DataFrame that does not define a constructor."""
    pass

class TestSubclassWithoutConstructor():

    def test_copy(self):
        expected = DataFrame({"a": [1, 2, 3]})
        result = SimpleSubClass(expected).copy()

        tm.assert_frame_equal(result, expected)

    def test_groupby(self):
        df = SimpleSubClass(DataFrame({"a": [1, 2, 3]}))

        for _, v in df.groupby("a"):
            assert isinstance(v, DataFrame)
```

added to `test_subclass.py`.

ivirshup committed:
Revert _constructor check

ivirshup committed:
Revert other constructor checks, allow warning in test_reindex_like_subclass

ivirshup committed:
Add tests

jbrockmendel commented:
cc @jorisvandenbossche im fine with the change here bc it improves perf the base classes.  no strong opinion on the actual behavior.

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -646,7 +646,7 @@ def _constructor(self) -> Callable[..., DataFrame]:
-        if self._constructor is DataFrame:
+        if type(self) is DataFrame:
             # we are pandas.DataFrame (or a subclass that doesn't override _constructor)
             return self._from_mgr(mgr, axes=axes)
```
Alternative fix:
```suggestion
            return DataFrame._from_mgr(mgr, axes=axes)
```

Since `_from_mgr` is a class method. 

That would avoid passing a manager object to the DataFrame(..) constructor. 
Or another alternative (but that makes it a bit more complicated), is to keep the `if self._constructor is DataFrame:` check in the else block, like:

```
...
else:
    if self._constructor is DataFrame:
        return DataFrame._from_mgr(mgr)
    else:
        return self._constructor(mgr) 

```

ivirshup replied:
I like the first approach you suggest here better since it's a class method. But also, this is in some internals I'm not very familiar with so happy to go with whatever other approach.

The performance benefit of my current approach seems small, and okay to go without, right?

jorisvandenbossche commented:
>  it improves perf the base classes

Do you mean that the check `type(self) is DataFrame` is a bit faster than `self._constructor is DataFrame`? (it's indeed a few nanoseconds faster)

ivirshup committed:
Merge branch 'main' into check-class-on-constructor-from-mgr

ivirshup committed:
Release note

ivirshup committed:
Add Series test, accept suggestion

ivirshup commented:
I've addressed the comment, added a test for subclasses of Series, and added a release note.

This does throw a depreciation warning:

```python
import pandas as pd

class SimpleDataFrameSubClass(pd.DataFrame):
    """A subclass of DataFrame that does not define a constructor."""

s = SimpleDataFrameSubClass(pd.DataFrame({"a": [1, 2, 3]}))
s.copy()
```

```
<ipython-input-2-6e1d2ee6d3ad>:1: DeprecationWarning: Passing a BlockManager to DataFrame is deprecated and will raise in a future version. Use public APIs instead.
  s.copy()
```

But [the previous fix seems to have done this as well](https://github.com/jorisvandenbossche/pandas/blob/0ffd772716969816388cff9597e5f5c941350458/pandas/tests/groupby/test_groupby_subclass.py#L14-L16).

It would make sense if this didn't warn, since the user code in my example above can't obviously do anything about this warning besides silence it. It may be out of scope for this regression though, since this PR does not introduce this warning.

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -646,9 +646,9 @@ def _constructor(self) -> Callable[..., DataFrame]:
 
     def _constructor_from_mgr(self, mgr, axes):
-        if self._constructor is DataFrame:
+        if type(self) is DataFrame:
```
Does the test also pass if you keep the original `self._constructor is DataFrame` here?

Because I think that should fix the deprecation warning

ivirshup replied:
It does. I think I can also revert the changes to 

`DataFrame. _constructor_sliced_from_mgr ` and `Series. _constructor_expanddim_from_mgr` in this case

jorisvandenbossche replied:
Yes, indeed

jorisvandenbossche commented:
Yes, we still have to figure out how to deal with this deprecation warning in subclasses. But, that deprecation warning is only for pandas 2.2, and so won't yet be visible in a pandas 2.1.3. 

And I think my other comment might solve this warning for the simple case where the subclass doesn't override the constructor.

jorisvandenbossche commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -773,3 +773,31 @@ def test_constructor_with_metadata():
+        expected = DataFrame({"a": [1, 2, 3]})
+        result = SimpleDataFrameSubClass(expected).copy()
+
+        tm.assert_frame_equal(result, expected)
```
Although we have a `check_frame_type` ("Whether to check the DataFrame class is identical.") that defaults to True, that seems to only check that the left frame is an instance of the class of the right frame. So that won't actually catch the issue if `result` was still a subclass instead of a DataFrame.

So in addition I would add an explicit:
```suggestion
        assert type(result) is DataFrame
        tm.assert_frame_equal(result, expected)
```

ivirshup replied:
Huh! Weird!

But I think this does explain some confusion I've had about this function. I've added a comment so this doesn't get removed by someone assuming the type check is covered.

```suggestion
        assert type(result) is DataFrame  # assert_frame_equal only checks isinstance(lhs, type(rhs))
        tm.assert_frame_equal(result, expected)
```

ivirshup replied:
`assert_series_equal` seems order invariant here

jorisvandenbossche replied:
We should probably call that a bug in `assert_frame_equal` ..

ivirshup committed:
Revert specific instance check, use of class method now catches this

ivirshup commented on pandas/core/frame.py:
```diff
@@ -659,7 +659,7 @@ def _sliced_from_mgr(self, mgr, axes) -> Series:
 
     def _constructor_sliced_from_mgr(self, mgr, axes):
-        if self._constructor_sliced is Series:
+        if type(self) is DataFrame:
```
```suggestion
        if self._constructor_sliced is Series:
```

ivirshup committed:
Fix type checks in tests

jorisvandenbossche committed:
Update pandas/core/frame.py

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

jorisvandenbossche committed:
Update pandas/core/series.py

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

jorisvandenbossche commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -773,3 +773,34 @@ def test_constructor_with_metadata():
+        result = SimpleSeriesSubClass(expected).copy()
+
+        tm.assert_series_equal(result, expected)
+
```
Could you add one test for `Series.to_frame()` as well? That covers `_constructor_expanddim_from_mgr`

ivirshup replied:
Added!

jorisvandenbossche reviewed (approved):
Assuming the tests pass, this looks good to me!

ivirshup committed:
Add test for Series.to_frame

jorisvandenbossche commented:
Thanks @ivirshup !
